pandas/io/parsers.py

```diff
@@ -923,14 +918,13 @@ def __init__(self, f, engine=None, **kwds):
         self.nrows = options.pop("nrows", None)
         self.squeeze = options.pop("squeeze", False)
 
-        # might mutate self.engine
-        self.engine = self._check_file_or_buffer(f, engine)
+        self._check_file_or_buffer(f, engine)
         self.options, self.engine = self._clean_options(options, engine)
 
         if "has_index_names" in kwds:
             self.options["has_index_names"] = kwds["has_index_names"]
 
-        self._make_engine(self.engine)
+        self._engine = self._make_engine(self.engine)
```

Comment:
Does assigning this back to self._engine buy us anything? IIUC would be better to just rename this method `_check_engine` and have it raise for an invalid one without returning anything

Reply:
This method actually does create ``_engine``. Here ``self.engine`` is a type of engine (string), while ``_engine`` is an actual engine (type ``ParserBase``). Originally this method did not return a value, but implicitly assigned to ``self._engine``. Meanwhile in ``__init__`` there was a declaration ``self._engine = None``.
I found this confusing. Why not just returning from ``self._make_engine`` the actual engine once? Setting up of the attributes is easier to see in ``__init__``, rather than in the body of some private methods. That was my intention to return from ``_make_engine``.

Reply:
Ah OK I see - thanks for clarifying